python/ray/data/dataset.py

```diff
@@ -2456,6 +2459,17 @@ def iter_batches(
                 to select ``numpy.ndarray`` for tensor datasets and
                 ``Dict[str, numpy.ndarray]`` for tabular datasets. Default is "native".
             drop_last: Whether to drop the last batch if it's incomplete.
+            shuffle: Whether to randomly shuffle the data using a local in-memory
+                shuffle buffer. This can only be used if a ``batch_size`` is specified.
+                This is a light-weight alternative to the global `.random_shuffle()`
+                operation; this shuffle will be less random but will be faster and less
+                resource-intensive.
+            shuffle_buffer_min_size: Minimum number of rows that must be in the local
```

Comment:
>the TF is an API on the Dataset and is a transformation not consumption/sink, it doesn't have two similar "*_size" parameters as well

>If they have such difference, why jam them together, i.e. would it be cleaner to decouple them into separate APIs?

TensorFlow Datasets has batching and shuffling (and prefetching) as separate operations; they can do this since TensorFlow Datasets are lazy by default. For Ray Datasets to support this combination efficiently with an eager-by-default execution model, these arguments need to be specified together, within the consumption API call. And since they are configuring how the data is going to be iterated over, I think this is reasonable until we move to a lazy-first execution model.

I should also note that this API has led to a lot of user confusion, bugs, and performance issues. E.g. batching before shuffling will batch and then shuffle the order of those batches (but not shuffling the rows within or between batches), which is rarely what the user wants (an Anyscale customer ran into this). If you batch upstream, you therefore have to unbatch, shuffle, and rebatch to get the desired shuffling semantics, which is very inefficient.

Consolidating those arguments to a single "convert the dataset to an iterator" API is more declarative, which allows us to:
- avoid these API call ordering bugs and user confusion
- make the shuffler batch-aware, which allows us to optimize for minimal data copying, better batch size + shuffle buffer size validation at API call time, etc.

>the Torch seems a sampler that may use shuffle underneath (not a good idea).

The sampler is a random access (index-based) sampler, so it's a shuffler that generates random indices and yields items with O(1) access for map-style datasets. The Torch example is a bit different, but it illustrates a common theme: users like to think about the size of the set out of which samples will be drawn in terms of samples, even if batching is taking place.

>The examples seem not quite the same as this API:

The Petastorm API is probably the most direct comparison, in which both the batch size and the shuffle queue capacity is specified in the same API, and the shuffle queue capacity is given in number of samples, not as a batch multiplier.

Reply:
And when configuring the shuffle buffer size in order to ensure good randomness in the shuffle, I think that ML and MLE users will prefer to think about it in terms of the number of samples rather than the number of batches.

From the user perspective, this API should be thought of as batching happening downstream from the shuffling (i.e. that the shuffling is independent from the batching); the fact that we use the batch size to implement a more efficient local shuffle is an implementation detail.

Reply:
You already said there are differences in TF/PT/Petastorm, doesn't it mean there is actually no "standard" API, and this is just following Petastorm?

We already have ds.shuffle() and then ds.iter_batches(). What makes it more confusing to have ds.local_shuffle() and then ds.iter_batches(), than entangling different things together?

Reply:
>You already said there are differences in TF/PT/Petastorm, doesn't it mean there is actually no "standard" API, and this is just following Petastorm?

The commonality that I'm highlighting there is that, for shuffle buffer implementations, the shuffle buffer size is given in the number of samples since that is how users like to configure the buffer when thinking about (1) the randomness of the shuffle and (2) the CPU memory constraint. Petastorm is the only one that implements batching + shuffling in the same API call, and yet they still give the shuffle buffer size in the number of samples, not as the number of batches, and for the aforementioned reasons.

Overall, I'm arguing that giving the shuffle buffer size in terms of the number of samples is the current standard and makes the most sense to our users. Do you still disagree with either?

>We already have ds.shuffle() and then ds.iter_batches(). What makes it more confusing to have ds.local_shuffle() and then ds.iter_batches(), than entangling different things together?

Because:

1. This is fundamentally a local, in-iterator operation, so having it as a separate stage doesn't make much sense (bad semantics for an independent operation, inefficient if eagerly executed) unless we can delay it to the iterator, which wouldn't work with eager execution semantics.
2. Making the shuffler batch-aware (and eventually prefetch-aware) has efficiency and performance benefits, hence Petastorm doing something similar.

There's already an intra-shard, lighter-weight shuffle option that's a standalone Dataset operation: shuffling after splitting.

```python
@ray.remote
def consume(split):
    for batch in split.random_shuffle_each_window().iter_batches():
        print("batch size:", len(batch))

splits = ray.data.range(100, parallelism=8).repeat(2).split(4, equal=True)

ray.get([consume.remote(split) for split in splits])
```

This will launch a smaller-scale, distributed, map-reduce shuffle only on the blocks within a shard.

Reply:
IMO the preference order of API will be:
- Decoupled without mixing the API responsibilities: ds.local_shuffle(shuffle_buffer_size).iter_batches(batch_size)
- Coupled API but with more consolidated semantics: ds.iter_batches(batch_size, num_batch_to_shuffle)
- Coupled API but with more separate semantics: ds.iter_batches(num_rows_in_batch, num_rows_to_shuffle)
If there is clear UX signal, we can reorder.
For performance, the current plan is to make it lazy by default anyway.